panoramapublic/src/org/labkey/panoramapublic/chromlib/ChromLibStateImporter.java

```diff
@@ -0,0 +1,406 @@
+{
+    protected final Logger _log;
+    protected final Container _container;
+    protected final User _user;
+    protected final TargetedMSService _svc;
+    protected ITargetedMSRun _currentRun;
+    private Map<LibPeptideGroup.LibPeptideGroupKey, Long> _pepGrpKeyMap;
+
+    abstract String libTypeString();
+    abstract List<String> getExpectedColumns();
+    abstract List<Header<?>> getHeaders();
+    abstract void updateRows();
+
+    public ChromLibStateImporter(Container container, User user, Logger log, TargetedMSService svc)
+    {
+        _container = container;
+        _user = user;
+        _log = log;
+        _svc = svc;
+    }
+
+    public static void importLibState(File libStateFile, Container container, User user, Logger log) throws ChromLibStateException
+    {
+        TargetedMSService svc = TargetedMSService.get();
+        TargetedMSService.FolderType folderType = svc.getFolderType(container);
+        if (TargetedMSService.FolderType.LibraryProtein.equals(folderType))
+        {
+            new ProteinLibStateImporter(container, user, log, svc).importFromFile(libStateFile);
+        }
+        else if (TargetedMSService.FolderType.Library.equals(folderType))
+        {
+            new PeptideLibStateImporter(container, user, log, svc).importFromFile(libStateFile);
+        }
+        else
+        {
+            throw new ChromLibStateException(String.format("'%s' is not a chromatogram library folder.", container));
+        }
+    }
+
+    void importFromFile(File libStateFile) throws ChromLibStateException
+    {
+        DbScope.Transaction transaction = PanoramaPublicSchema.getSchema().getScope().getCurrentTransaction();
+        if (transaction == null)
+        {
+            throw new IllegalStateException("Callers should start their own transaction");
+        }
+
+        _log.info(String.format("Importing %s library state in container '%s' to file '%s'.", libTypeString(), _container.getPath(), libStateFile.getPath()));
+
+        try (TabLoader reader = new TabLoader(libStateFile, true))
+        {
+            CloseableIterator<Map<String, Object>> iterator = reader.iterator();
+            try
+            {
+                // Verify that we have the right column headers
+                var columns = reader.getColumns();
+                verifyLibColumns(columns, getExpectedColumns());
+                setColumnTypes(columns);
+            }
+            catch (IOException e)
+            {
+                throw new ChromLibStateException(String.format("Error reading header columns from file '%s'.", libStateFile));
+            }
+            while(iterator.hasNext())
+            {
+                parseLibStateRow(iterator.next());
+            }
+        }
+        updateRows();
+        _log.info("Done importing library state.");
+    }
+
+    private void verifyLibColumns(ColumnDescriptor[] columns, List<String> expectedColulmns) throws ChromLibStateException
```

Comment:
```suggestion
    private void verifyLibColumns(ColumnDescriptor[] columns, List<String> expectedColumns) throws ChromLibStateException
```

(would need typo corrections in body below too)